client/finality-grandpa/src/communication/gossip.rs

```diff
@@ -469,14 +469,17 @@ impl<N> Default for Peers<N> {
 }
 
 impl<N: Ord> Peers<N> {
-	fn new_peer(&mut self, who: PeerId, roles: Roles) {
-		if roles.is_authority() && self.lucky_authorities.len() < MIN_LUCKY {
-			self.lucky_authorities.insert(who.clone());
-		}
-		if !roles.is_authority() && self.lucky_peers.len() < MIN_LUCKY {
-			self.lucky_peers.insert(who.clone());
+	fn new_peer(&mut self, who: PeerId, role: ObservedRole) {
+		match role {
+			ObservedRole::Authority if self.lucky_authorities.len() < MIN_LUCKY => {
+				self.lucky_authorities.insert(who.clone());
```

Comment:
We should keep them in a separate vec, both `lucky_peers` and `lucky_authorities` get reshuffled and truncated after every round. We want to make sure our sentries and guarded authorities are always selectd.

Reply:
As far as I understand, `lucky_peers` and `lucky_authorities` are only used to determine who to send messages to, and the code that actually determines that has a short circuit that always returns `true` for `OurSentry` and `OurGuardedAuthorities`, which is why I don't include them.

Reply:
https://github.com/paritytech/substrate/blob/5b5d5d0ea5fdad18dae55f3942e1fada711f5dcf/client/finality-grandpa/src/communication/gossip.rs#L575

Reply:
The lucky peers and lucky authorities are a subset of peers that are randomly selected at each round to prioritize sending messages to, i.e. the lucky ones change every round. Currently our sentry nodes and guarded authorities are added to this subset, but they may be removed the round after.

Reply:
Sorry for the noise, I'm an idiot. For some reason when I reviewed this I thought we were adding the sentries and guarded validators to the lucky peer set. :man_facepalming: